src/log/log_slot.c

```diff
@@ -370,53 +479,65 @@ __wt_log_slot_join(WT_SESSION_IMPL *session, uint64_t mysize,
 	WT_LOGSLOT *slot;
 	int64_t flag_state, new_state, old_state, released;
 	int32_t join_offset, new_join;
+	bool unbuffered;
 #ifdef	HAVE_DIAGNOSTIC
-	bool unbuf_force;
+	bool yld;
 #endif
 
 	conn = S2C(session);
 	log = conn->log;
 
 	WT_ASSERT(session, !F_ISSET(session, WT_SESSION_LOCKED_SLOT));
+	WT_ASSERT(session, mysize != 0);
 
 	/*
 	 * There should almost always be a slot open.
 	 */
+	unbuffered = false;
 #ifdef	HAVE_DIAGNOSTIC
-	unbuf_force = (++log->write_calls % WT_THOUSAND) == 0;
+	yld = (log->write_calls % 7) == 0;
+	if ((++log->write_calls % WT_THOUSAND) == 0 ||
+	    mysize > WT_LOG_SLOT_BUF_MAX) {
+#else
+	if (mysize > WT_LOG_SLOT_BUF_MAX) {
 #endif
+		unbuffered = true;
+		F_SET(myslot, WT_MYSLOT_UNBUFFERED);
+	}
 	for (;;) {
 		WT_BARRIER();
-		slot = log->active_slot;
-		old_state = slot->slot_state;
-		/*
-		 * Try to join our size into the existing size and
-		 * atomically write it back into the state.
-		 */
-		flag_state = WT_LOG_SLOT_FLAGS(old_state);
-		released = WT_LOG_SLOT_RELEASED(old_state);
-		join_offset = WT_LOG_SLOT_JOINED(old_state);
+		old_state = log->active_slot->slot_state;
+		if (WT_LOG_SLOT_OPEN(old_state)) {
+			/*
+			 * Try to join our size into the existing size and
+			 * atomically write it back into the state.
+			 */
+			flag_state = WT_LOG_SLOT_FLAGS(old_state);
+			released = WT_LOG_SLOT_RELEASED(old_state);
+			join_offset = WT_LOG_SLOT_JOINED(old_state);
+			if (unbuffered)
+				new_join = join_offset + WT_LOG_SLOT_UNBUFFERED;
+			else
+				new_join = join_offset + (int32_t)mysize;
+			new_state = (int64_t)WT_LOG_SLOT_JOIN_REL(
+			    (int64_t)new_join, (int64_t)released,
+			    (int64_t)flag_state);
+
+			slot = log->active_slot;
```

Comment:
Is there a race possible here when reading `log->active_lot`?  In other words, does it matter if `slot` ends up with a different location than we read `old_state` from above?